lib/lseek64_stubs.c

```diff
@@ -46,12 +46,10 @@ CAMLprim value stub_lseek64_data(value fd, value ofs) {
   off_t c_ret;
 
   caml_enter_blocking_section();
-#if defined(SEEK_DATA)
   c_ret = lseek(c_fd, c_ofs, SEEK_DATA);
-#else
-  /* Set the file pointer to ofs; pretend there is data */
-  c_ret = lseek(c_fd, c_ofs, SEEK_SET);
-#endif
+  /* If there is no SEEK_DATA support, 
+     set the file pointer to ofs; pretend there is data */
+  if (c_ret == -1) c_ret = lseek(c_fd, c_ofs, SEEK_SET);
```

Comment:
Which is why you have to then call the fallback, to tell if the issue is
SEEK_DATA not being implemented, or one of the other EINVAL results.

I don't see how, with EINVAL being overloaded, to do this detection at
runtime without calling the fallback, too.

On Thu, May 26, 2016 at 5:24 AM Euan Harris <notifications@github.com>
wrote:

> In lib/lseek64_stubs.c
> <https://github.com/djs55/ocaml-vhd/pull/33#discussion_r64723270>:
>
> >    c_ret = lseek(c_fd, c_ofs, SEEK_DATA);
> > -#else
> > -  /* Set the file pointer to ofs; pretend there is data */
> > -  c_ret = lseek(c_fd, c_ofs, SEEK_SET);
> > -#endif
> > +  /* If there is no SEEK_DATA support,
> > +     set the file pointer to ofs; pretend there is data */
> > +  if (c_ret == -1) c_ret = lseek(c_fd, c_ofs, SEEK_SET);
>
> lseek should return EINVAL if whence (in this case SEEK_DATA) is invalid.
> Unfortunately this error code is also returned for other error conditions:
>
> ERRORS
>        EINVAL whence is not valid.  Or: the resulting file offset would be  negative,  or
>               beyond the end of a seekable device.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> <https://github.com/djs55/ocaml-vhd/pull/33/files/476f501e0d75dfba04750f3f33135e648755a5c6#r64723270>
>